src/SonarQube.Client/Services/SonarQubeClient.cs

```diff
@@ -0,0 +1,264 @@
+
+                    return new CredentialsDTO { AreValid = isValid };
+                });
+        }
+
+        internal /*for testing purposes*/ static string AppendQueryString(string urlBase, string queryFormat,
+            params object[] args)
+        {
+            return urlBase +
+                string.Format(CultureInfo.InvariantCulture, queryFormat,
+                    args.Select(x => HttpUtility.UrlEncode(x.ToString())).ToArray());
+        }
+
+        internal /*for testing purposes*/ static Uri CreateRequestUrl(HttpClient client, string apiUrl)
+        {
+            // We normalize these inputs to that the client base address always has a trailing slash,
+            // and the API URL has no leading slash.
+            // Failure to do so will cause an incorrect URL to be formed, with the apiUrl being relative
+            // to the base address's hostname only.
+            Debug.Assert(client.BaseAddress.ToString().EndsWith("/"),
+                "HttpClient.BaseAddress should have a trailing slash");
+            Debug.Assert(!apiUrl.StartsWith("/"),
+                "API URLs should not begin with a slash as this forces the API to be relative to the base URI's host.");
+
+            Uri normalBaseUri = client.BaseAddress.EnsureTrailingSlash();
+            string normalApiUrl = apiUrl.TrimStart('/');
+
+            return new Uri(normalBaseUri, normalApiUrl);
+        }
+
+        private static async Task<string> GetStringResultAsync(HttpResponseMessage response,
+            CancellationToken token)
+        {
+            token.ThrowIfCancellationRequested();
+            response.EnsureSuccessStatusCode();
+
+            return await response.Content.ReadAsStringAsync().ConfigureAwait(false);
+        }
+
+        private static async Task<HttpResponseMessage> InvokeGetRequest(HttpClient client, string apiUrl,
+                    CancellationToken token)
+        {
+            var uri = CreateRequestUrl(client, apiUrl);
+            var httpRequest = new HttpRequestMessage(HttpMethod.Get, uri);
+            var httpResponse = await client.SendAsync(httpRequest, HttpCompletionOption.ResponseHeadersRead, token)
+                .ConfigureAwait(false);
+
+            return httpResponse;
+        }
+        private static T ProcessJsonResponse<T>(string jsonResponse, CancellationToken token)
+        {
+            token.ThrowIfCancellationRequested();
+            return JsonHelper.Deserialize<T>(jsonResponse);
+        }
+
+        private async Task<Result<T>> SafeUseHttpClient<T>(ConnectionDTO connection, Func<HttpClient, Task<T>> sendRequestsAsync)
+        {
+            using (HttpClient client = new HttpClient(this.messageHandler))
+            {
+                client.BaseAddress = connection.ServerUri;
+                client.DefaultRequestHeaders.Authorization =
+                    AuthenticationHeaderHelper.GetAuthenticationHeader(connection);
```

Comment:
This would read nicer as something along the lines of `AuthenticationHeader.BuildFrom(connection)`